### PR TITLE
Improve memory usage when creating cluster messages

### LINKED ISSUE #6476: Memory usage in cluster messaging
|Wazuh version|Component|Install type|
|---|---|---|
|4.0| Wazuh cluster | Manager |

## Description

Hey team,

We need to improve the part of the code in charge of creating a message to be sent in the cluster.

### Fix ram usage for bytearray variable
The actual cluster code is using a fixed 10 Mb bytearray for carrying out message builds. This is improper because the message might be just a few bytes and we are always using 10 Mb of memory. Also, when several connections are made, we can have multiple of these variables stored in the memory, greatly increasing memory usage.

https://github.com/wazuh/wazuh/blob/master/framework/wazuh/core/cluster/common.py#L159-L162

https://github.com/wazuh/wazuh/blob/master/framework/wazuh/core/cluster/common.py#L206-L210

An already applied fix (never got merged) can be found in https://github.com/wazuh/wazuh/pull/4481.

Pympler usage can be found in https://github.com/wazuh/wazuh/issues/4444 and https://github.com/wazuh/wazuh/issues/4463.

## To do
- [x] Learn to use tools to understand python memory allocation (pympler) and find possible memory leaks
- [x] Improve memory usage when building cluster messages

Best regards,

David J. Iglesias

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #6476 |

Hi team,

This PR closes #6476. I have learnt to use `pympler` and find possible memory leaks. In the related issue I posted a Python example of use of `pympler` with a memory leak example (using a mutable type as a default parameter in a function).

I have improved the memory usage when building cluster messages. The cluster code was using a 10 MB `bitearray` for carrying out message builds. This led to memory leaks when this `bitearray` was using less than his size (almost always).

With my changes, the message is more dynamic and we don't have any memory leaks in the cluster message creation.

The following Python script shows the old memory allocation and the new one when creating a cluster message in a summary table:

```
import struct
import random

from pympler import tracker

# variables used for the message creation
data = b'data'
command = b'command'
counter = random.SystemRandom().randint(0, 2 ** 32 - 1)
cmd_len = 12
header_len = cmd_len + 8
request_chunk = 5242880
header_format = '!2I{}s'.format(cmd_len)

# initialize summary tracker
memory_tracker = tracker.SummaryTracker()

# Initial summaries
print("\nInitial summaries")
memory_tracker.print_diff()
memory_tracker.print_diff()


# Memory leak
out_msg = bytearray(header_len + request_chunk * 2)
out_msg[:header_len] = struct.pack(header_format, counter, len(data), command)
out_msg[header_len:header_len + len(data)] = data

# Actual message size
#       types |   # objects |   total size
# =========== | =========== | ============
#   bytearray |           1 |     10.00 MB

print("\nActual message object")
memory_tracker.print_diff()
print("\nMake the diff empty")
memory_tracker.print_diff()


# NO Memory leak, difference with the last message
out_msg = bytearray(header_len + len(data))
out_msg[:header_len] = struct.pack(header_format, counter, len(data), command)
out_msg[header_len:header_len + len(data)] = data

# Message size difference. It will depend on the msg sent
#       types |   # objects |        total size
# =========== | =========== | =================
#   bytearray |           0 |   -10485756     B

print("\nNew message object")
memory_tracker.print_diff()
```

As we can see, the old message size was 10 MB (=10485760). With `pympler` we can see the difference when recreating the message in a dynamic way (adding the data length instead of introducing the data in a static array with a bigger size). The difference for this example is -10485756 B, which means the new message size is 10485760-10485756 = `4 B`.

- Unit test results:

```
framework/wazuh/tests/test_cluster.py .......                            [100%]
============================== 7 passed in 0.11s ===============================
```

```
framework/wazuh/core/cluster/tests/test_cluster.py ...................   [ 34%]
framework/wazuh/core/cluster/tests/test_common.py .......                [ 47%]
framework/wazuh/core/cluster/tests/test_control.py .....                 [ 56%]
framework/wazuh/core/cluster/tests/test_local_client.py .                [ 58%]
framework/wazuh/core/cluster/tests/test_utils.py .......                 [ 70%]
framework/wazuh/core/cluster/tests/test_worker.py .........FF.....       [100%]
Only 2 failed tests in test_worker.py, non related to the changes I did
```

```
framework/wazuh/core/cluster/dapi/tests/test_dapi.py
======================= 22 passed, 11 warnings in 0.83s ========================
```

Regards,
Manuel.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 - Fixed a `cluster_control` bug that caused an error message when running `wazuh-clusterd` in foreground. ([#6724](https://github.com/wazuh/wazuh/pull/6724))
+- Avoid memory leaks when creating cluster messages. ([#6736](https://github.com/wazuh/wazuh/pull/6736))
```
Comment (davidjiglesias) [inline CHANGELOG.md:46]:
```suggestion
- Improved memory usage when creating cluster messages. ([#6736](https://github.com/wazuh/wazuh/pull/6736))
```

**Review by davidjiglesias — CHANGES_REQUESTED** [review]
Review code changes

**Comment by Selutario** [comment]
## Issue update
Hello team!

There are some things to keep in mind after the changes made for this issue, which could lead to new bugs. On the other hand, there are possible bugs that have been in the code for a long time and that we have discovered thanks to the changes applied here. I detail them below.

### 1. Long encrypted data length
Before the changes, as the space was reserved for the `out_msg` bytearray in advance, if more space was used than reserved (which could happen after encrypting it in `self.msg_build`), a MemoryError exception could be generated as seen in the 355-357 lines. 

https://github.com/wazuh/wazuh/blob/1acc9efbcf3f7be5e17e90248b68b3c50ca9fd9b/framework/wazuh/core/cluster/common.py#L332-L359

Now that exception seems not to occur, although it is still being handled. An exception is still thrown if the unencrypted data exceeds `self.request_chunk` bytes. However, if once encrypted the data do exceed it, the MemoryError exception will not be raised inside the `self.msg_build`. 

https://github.com/wazuh/wazuh/blob/1acc9efbcf3f7be5e17e90248b68b3c50ca9fd9b/framework/wazuh/core/cluster/common.py#L245-L276

**TODO:** investigate and reproduce what is happening in these cases, and modify `msg_build` to take into account the size after encryption and perform the proper actions if exceeded.

```Python
        # adds - to command until it reaches cmd length
        command = command + b' ' + b'-' * (self.cmd_len - cmd_len - 1)
        encrypted_data = self.my_fernet.encrypt(data) if self.my_fernet is not None else data
        if len(encrypted_data) <= self.request_chunk*2:
            out_msg = bytearray(self.header_len + len(encrypted_data))
            out_msg[:self.header_len] = struct.pack(self.header_format, counter, len(encrypted_data), command)
            out_msg[self.header_len:self.header_len + len(encrypted_data)] = encrypted_data
        else:
            # Study and perform best action: I.e: raise MemoryError? Try to find a better option
```

### 2. dispatch() not handling msg_build exception
As seen before, `msg_build()` method could raise either a WazuhClusterError (Line 267) or, at least until now, a MemoryError. It is handled inside send_request() method, but `msg_build()` is also called in `dispatch()` which is not handling it (line 527), nor in upper instances, so it could kill the `wazuh-clusterd` process.

https://github.com/wazuh/wazuh/blob/1acc9efbcf3f7be5e17e90248b68b3c50ca9fd9b/framework/wazuh/core/cluster/common.py#L505-L527

**TODO:** Handle possible exceptions in a similar way to the rest of the `dispatch()` method.

### 3. DAPI and sendsync long data`
send_request()` method is being used in many places. However, although it could raise a WazuhClusterError, no precautions are being taken in those places where, for instance, the `DAPI` command or `sendsync` is used. An example is this:
https://github.com/wazuh/wazuh/blob/1acc9efbcf3f7be5e17e90248b68b3c50ca9fd9b/framework/wazuh/core/cluster/local_server.py#L371

If `data` is longer than `self.request_chunk`, the process will raise a (handled) exception and it will never be completed, as no other actions are taken in order to fix the problem (send half of data, etc). 

**TODO:** Investigate and reproduce the problem, find possible solutions and fix all those places that may be failing right now.

Many of these tasks are broad enough to open a well-detailed new issue and work individually. Feel free to do it.

Best regards,
Selu.

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: framework/wazuh/core/cluster/common.py [hunk 2]
```diff
@@ -271,10 +269,11 @@ def msg_build(self, command: bytes, counter: int, data: bytes) -> bytes:
         # adds - to command until it reaches cmd length
         command = command + b' ' + b'-' * (self.cmd_len - cmd_len - 1)
         encrypted_data = self.my_fernet.encrypt(data) if self.my_fernet is not None else data
-        self.out_msg[:self.header_len] = struct.pack(self.header_format, counter, len(encrypted_data), command)
-        self.out_msg[self.header_len:self.header_len + len(encrypted_data)] = encrypted_data
+        out_msg = bytearray(self.header_len + len(encrypted_data))
```
[the remark this answers is not in the document]
Comment (mcarmona99) [reply framework/wazuh/core/cluster/common.py:272]: It is not neccessary, it wasnt checked without the changes.
The check will be added in the refactor done in https://github.com/wazuh/wazuh/issues/6820

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**